### PR TITLE
Add scope_exit helper.

### PR DESCRIPTION
**Pull request by hidmic**
Based off of `rclcpp`'s. Can be cancelled.

### CODE REVIEW

File: test/test_scope_exit.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+      [&called]() {
+        called = true;
+      });
+    on_exit.cancel();
+  }
+
+  EXPECT_FALSE(called);
+}
+
+TEST(test_scope_exit, code_types) {
```
Comment (hidmic) [inline test/test_scope_exit.cpp:49]: If anybody has a nice way to "test" signatures in compile time (w/o an actual bare test), I'm all ears.

**Comment by hidmic** [comment]
CI up to `rcpputils`:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11295/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6542/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9245/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11199/)

**Comment by hidmic** [comment]
> lgtm, we should remove the one in rclcpp and use this instead.

Will do, but I want the original `rmw` PRs in first :sweat_smile: !

File: include/rcpputils/macros.hpp [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCPPUTILS__MACROS_HPP_
+#define RCPPUTILS__MACROS_HPP_
+
+#define RCPPUTILS_STRING_JOIN(arg1, arg2) RCPPUTILS_DO_STRING_JOIN(arg1, arg2)
+#define RCPPUTILS_DO_STRING_JOIN(arg1, arg2) arg1 ## arg2
```
Comment (ivanpauno) [inline include/rcpputils/macros.hpp:19]: this could live in `rcutils` `macros.h`, but it doesn't matter much

File: test/test_scope_exit.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  rcpputils::make_scope_exit(const_code);
+
+  struct noncopyable_code
+  {
+    noncopyable_code() = default;
+    noncopyable_code(const noncopyable_code &) = delete;
+    noncopyable_code(noncopyable_code &&) = default;
+    void operator()() {}
+  };
+  rcpputils::make_scope_exit(noncopyable_code());
```
Comment (ivanpauno) [inline test/test_scope_exit.cpp:63]:
```suggestion
  rcpputils::make_scope_exit(noncopyable_code);
```
??

Comment (hidmic) [reply test/test_scope_exit.cpp:63]: Hmm, `noncopyable_code` is a type. Maybe a bit misleading.

File: include/rcpputils/macros.hpp (hunk 2 above)
Comment (hidmic) [reply include/rcpputils/macros.hpp:19]: That's annoying. But I agree.

File: test/test_scope_exit.cpp (hunk 3 above)
Comment (ivanpauno) [reply test/test_scope_exit.cpp:63]: Sorry, my bad (I thought you already instantiated the object, and were calling the `operator()`).
I would use CamelCase per the style guide.

Comment (hidmic) [reply test/test_scope_exit.cpp]: See https://github.com/ros2/rcpputils/pull/78/commits/1360bc9071bda3eda708c31e91e029bf3b8dbfe5.

Comment (ivanpauno) [reply test/test_scope_exit.cpp]: From https://index.ros.org/doc/ros2/Contributing/Code-Style-Language-Versions/#function-and-method-naming:

> Note that classes should still use CamelCase by default

File: include/rcpputils/scope_exit.hpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+scope_exit<CallableT>
+make_scope_exit(CallableT && callable)
+{
+  return scope_exit<CallableT>(std::forward<CallableT>(callable));
+}
+
+}  // namespace rcpputils
+
+#define RCPPUTILS_SCOPE_EXIT(code) \
+  auto RCUTILS_JOIN(scope_exit_, __LINE__) = rcpputils::make_scope_exit([&]() {code;})
```
Comment (fujitatomoya) [inline include/rcpputils/scope_exit.hpp:66]: RCUTILS_JOIN will be defined in `rcutils`, right? I was not able to find it. just want to confirm.

Comment (hidmic) [reply include/rcpputils/scope_exit.hpp:66]: @fujitatomoya see https://github.com/ros2/rcutils/commit/adf1880e16edb97984b51784e1939c76050ed650.

**Comment by hidmic** [comment]
@ros-pull-request-builder retest this please

**Comment by hidmic** [comment]
Rolling PR job passed http://build.ros2.org/view/Rpr/job/Rbin_uF64__rcpputils__ubuntu_focal_amd64__binary/5/

**Pull request merged**